src/renderers/dom/fiber/ReactDOMFiber.js

```diff
@@ -359,6 +364,65 @@ var DOMRenderer = ReactFiberReconciler({
     parentInstance.removeChild(child);
   },
 
+  canHydrateInstance(
+    instance: Instance | TextInstance,
+    type: string,
+    props: Props,
+  ): boolean {
+    return instance.nodeType === 1 && type === instance.nodeName.toLowerCase();
+  },
+
+  canHydrateTextInstance(instance: Instance | TextInstance): boolean {
+    return instance.nodeType === 3;
+  },
+
+  getNextHydratableSibling(
+    instance: Instance | TextInstance,
+  ): null | Instance | TextInstance {
+    let node = instance.nextSibling;
+    // Skip non-hydratable nodes.
+    /*
+    while (node && node.nodeType !== 1 && node.nodeType !== 3) {
+      node = node.nextSibling;
+    }
+    */
+    return (node: any);
+  },
+
+  getFirstHydratableChild(
+    parentInstance: Container | Instance,
+  ): null | Instance | TextInstance {
+    let next = parentInstance.firstChild;
+    // Skip non-hydratable nodes.
+    /*
+    while (next && next.nodeType !== 1 && next.nodeType !== 3) {
+      next = next.nextSibling;
+    }
+    */
+    return (next: any);
+  },
+
+  hydrateInstance(
+    instance: Instance,
+    type: string,
+    props: Props,
+    rootContainerInstance: Container,
+    internalInstanceHandle: Object,
+  ): void {
+    precacheFiberNode(internalInstanceHandle, instance);
+    // TODO: Possibly defer this until the commit phase where all the events
+    // get attached.
+    updateFiberProps(instance, props);
+    setInitialProperties(instance, type, props, rootContainerInstance);
```

Comment:
I don’t want to conflate this issue too deeply with this tangent knowing it’s a non-goal so let me know if it makes sense to have a tracking issue for thoughts/best practices (whether that turns into implementation or documentation in the form section)

> We could potentially always fire a change event after reviving with the new value.
> It seems to me that most use cases for this would suffer in some way unless you designed both experiences.

I’m curious what use cases would suffer or hurt the user experience with this? I’ve experienced this scenario from browser prefilling forms from autocomplete setting and resetting the value during a refresh.

In both scenarios having it modeled as an onmount `setState` would’ve been sufficient to hydrate the host environment values back to React component state.